src/msix/pack/AppxPackageWriter.cpp

```diff
@@ -64,7 +110,77 @@ namespace MSIX {
+            files.push_back(std::move(payloadFile));
+        }
+        ProcessPayloadFiles(files);
+
         NOTIMPLEMENTED
     } CATCH_RETURN();
-}
\ No newline at end of file
+
+    std::unique_ptr<PayloadFile> AppxPackageWriter::BuildPayloadFile(const std::string& name, const ComPtr<IStream>& stream,
+        const std::string& contentType, APPX_COMPRESSION_OPTION compressionOpt)
+    {
+        // This might be called with external IStream implementations. Don't rely on internal implementation of FileStream
+        std::unique_ptr<PayloadFile> payloadFile = std::make_unique<PayloadFile>();
+        payloadFile->relativeName = name;
+        payloadFile->compressionOption = compressionOpt;
+
+        // Add content type to [Content Types].xml
+        m_contentTypeWriter->AddContentType(name, contentType);
+
+        LARGE_INTEGER start = { 0 };
+        ULARGE_INTEGER end = { 0 };
+        ThrowHrIfFailed(stream->Seek(start, StreamBase::Reference::END, &end));
+        ThrowHrIfFailed(stream->Seek(start, StreamBase::Reference::START, nullptr));
+        payloadFile->fileSize = static_cast<std::uint64_t>(end.u.LowPart);
+
+        std::unique_ptr<BlockAndHash> blockData;
+        std::uint64_t bytesToRead = payloadFile->fileSize;
+
+        while (bytesToRead > 0)
+        {
+            // Calculate the size of the next block to add
+            std::uint32_t blockSize = (bytesToRead > defaultBlockSize) ? defaultBlockSize : static_cast<std::uint32_t>(bytesToRead);
+            bytesToRead -= blockSize;
+
+            std::vector<std::uint8_t> buffer;
+            buffer.resize(blockSize);
+            ULONG bytesRead;
+            ThrowHrIfFailed(stream->Read(static_cast<void*>(buffer.data()), static_cast<ULONG>(blockSize), &bytesRead));
+            ThrowErrorIfNot(Error::FileRead, (static_cast<ULONG>(blockSize) == bytesRead), "Read stream file failed");
+
+            blockData.reset(new (std::nothrow) BlockAndHash());
+            blockData->block = std::move(buffer);
+            payloadFile->fileBlocks.push_back(std::move(blockData));
+        }
+        return payloadFile;
+    }
+
+    void AppxPackageWriter::ProcessPayloadFiles(const std::vector<std::unique_ptr<PayloadFile>>& files)
+    {
+        // calculate hash on parallel
+        auto computeHash = [](std::vector<std::uint8_t>& block) -> std::vector<std::uint8_t>
+        {
+            std::vector<std::uint8_t> hash;
+            ThrowErrorIfNot(Error::Unexpected, 
+                SHA256::ComputeHash(block.data(), static_cast<std::uint32_t>(block.size()), hash),
+                "Failed computing SHA256");
+            return hash;
+        };
+        for(auto& file : files)
+        {
+            for(auto& block : file->fileBlocks)
+            {
+                auto bind = std::bind(computeHash, block->block);
+                block->hashValue = std::async(std::launch::async, bind);
+            }
+        }

```

Comment:
This isn't what we need to parallelize.  And we shouldn't worry about parallel implementation just yet.  Try to design to allow it, but don't need to do it.